README.md
# **mmpack** - a user package manager

**mmpack** is a cross-platform, multi-versioning user package manager (as
opposed to system package manager). It leverages the container namespace
virtualisation technique to isolate package installations which allows
separate, non-conflicting ecosystems.

In the mmpack jargon, these ecosystems are called _prefixes_ and ressemble the
equivalent Python virtualenvs or Ruby gemsets.  However, mmpack packages are
not targeting any specific language, instead, they aim to bundle any meaningful
set of files (including binaries, data and documentation).

mmpack is designed to work without any need for root access, and to allow
multiple coexisting project versions within project prefixes.

mmpack leverages [mmlib](https://github.com/mmlabs-mindmaze/mmlib) and it is
hence fully cross-platform. Currently supported systems are GNU/Linux and
Windows.

For questions, requests, issues, pull-requests, etc., drop an email:
`nicolas.bourdaud@gmail.com`

## Main features

* Easy package creation
* User space installation
* Coexistence of different versions of the same software
* Deployment support (quick process to setup cross-platform packages)
* Built-in package distribution utility

## Installation

**mmpack** depends on [*mmlib*](https://github.com/mmlabs-mindmaze/mmlib),
*libcurl*, *libarchive* and *libyaml* and the python3 modules: *argcomplete*,
*astroid*, *elftools*, *pefile*, *urllib3*, *watchdog*, *yaml*.

After downloading the code, change to the source folder and type the following
to install mmpack:

``` bash
$> meson build && cd build
$> meson install
```

In GNU/Linux system also the _setcap_ command is required (with root
permission);

``` bash
#> ninja setcap
```

That allows mmpack to use the kernel namespace isolation for the prefixes and
it is the last time root permission is ever needed while using mmpack.

#### Debian systems

Debian users can take advantaged from a packaged version of mmpack. In debian
systems just add the file _/etc/apt/sources.list.d/mmpack.list_, containing:

```
deb http://opensource.mindmaze.com/debrepos unstable main
deb-src http://opensource.mindmaze.com/debrepos unstable main
```

The gpg key must also be added in the apt gpg folder:

``` bash
curl https://opensource.mindmaze.com/debrepos/mindmaze-opensource.gpg | sudo tee /etc/apt/trusted.gpg.d/mindmaze-opensource.gpg > /dev/null
```

Then it is a matter of (as root or with sudo):

``` bash
apt update
apt install mmpack
```

#### Windows systems (MSYS2)

Windows users with the MSYS2 system can edit _/etc/pacman.conf_ and add the
following lines:

```
[mmlabs]
Server = https://opensource.mindmaze.com/mingw
```

Then run:

``` bash
pacman -Sy
pacman -S mmpack
```

## Usage

mmpack comes with two CLI commands;

 * mmpack: manage prefixes and install software;
 * mmpack-build: create packages.

#### On prefixes

A prefix is identified by a path which can be specified in different ways;
through the command line argument --prefix and through the environment variable
$MMPACK_PREFIX.

Each prefix can be associated with one or several upstream repositories for
installing software right away; to create a prefix run:

``` bash
mmpack mkprefix --url=<http://repository/url> $MMPACK_PREFIX
```

(See the section below about installing software through repositories)

Process run outside the prefixes do not have access to their content and
processes run inside one prefix have the privileges of the running user plus
access to their own prefix (but not the others).

This means that, for a given software A with dependencies B,C, it is possible
to have:

 * a system wide installation at version x, with B,C
 * another installation at version y in prefix _j_
 * another installation at version z in prefix _k_, with patched B

Moreover, installation in a prefix does not require root privileges (provided
it is inside a user accessible folder).

#### Installing upstream software

mmpack sports some of the most common commands among package managers:

``` bash
# update package list from the repository url(s) of prefix <prefix_name>
mmpack --prefix=<prefix_name> update

# search for available package in the prefix <prefix_name>
mmpack --prefix=<prefix_name> search <package-name>

# install <package-name> in the prefix <prefix_name>
mmpack --prefix=<prefix_name> install <package-name>
```

Other useful commands include _list_ and _upgrade_.
All the commands support the _--help_ flag for discoverability.

All the commands provide the _--prefix_ flag, however it is also possible to
use the environment variable $MMPACK_PREFIX; the following commands have hence
the same effect as the above ones:

``` bash
export $MMPACK_PREFIX=<prefix_path>

# update package list from the repository url(s) of prefix <prefix_path>
mmpack update

# search for available package in the prefix <prefix_path>
mmpack search <package-name>

# install <package-name> in the prefix <prefix_path>
mmpack install <package-name>
```

#### Run mmpack software

To run mmack software, just invoke the `<command>` with:

``` bash
mmpack run <command>
```

Referring to the example for software A:

``` bash
A #  runs the system-wide installed version
mmpack --prefix=j run A #  runs A from prefix j
mmpack --prefix=k run A #  runs A from prefix k, which loads patched B
```

If no command is provided, the default $SHELL is launched.  Inside this shell
all the programs installed in the current prefix are added to the $PATH.
(Hence, typing _mmpack run_ is to some extent similar to mounting an extension
to the root filesystem).

Again, with respect to the above example,

``` bash
mmpack --prefix=j run #  enters prefix j environment
A #  runs A from prefix j
```

For the ease of use, the possibility to specify prefixes with an environment
variable has been introduced; so the following two invocations have the same
effect;

``` bash
mmpack --prefix=j run A #  runs A from prefix j

export MMPACK_PREFIX=j
mmpack run A #  runs A from prefix j
```

#### Package creation

Creating mmpack packages has been designed to be easy-peasy.  mmpack-build
supports the most widespread build-systems: Makefile, Autotools, Meson.

mmpack-build just needs the so-called _mmpack/spec_-file; a simple example is
given by

``` yaml
name: mmpack-hello-world
version: 1.0.0
maintainer: Gandalf <gandalf@the.grey>
url: ssh://<your git url>
description: mmpack hello world
```

The build-system to use can be specified; if it is not, mmpack will guess it
for you.

Refer to the [documentation](https://opensource.mindmaze.com/projects/mmpack/)
for in-depth details about spec files.  Provided sources and the mmpack/spec
file, building the package is as easy as typing:

``` bash
mmpack-build pkg-create
```

## Documentation

Documentation can be found online, https://opensource.mindmaze.com/projects/mmpack/ or
generated from sources.


debian/changelog
mmpack (0.7.9) unstable; urgency=medium

  * Hidden files are now properly listed in sha256sums file of packages.
  * validate file integrity upon mpk unpacking (disabled with
    MMPACK_DISABLE_CHECK_MPK_INTEGRITY env var).

 -- Nicolas Bourdaud <nicolas.bourdaud@gmail.com>  Thu, 02 Feb 2023 06:31:06 +0100

mmpack (0.7.8) unstable; urgency=medium

  * Use nettle library for computing SHA256.
  * packages can be installed from other prefixes where they are already
    installed (disabled with MMPACK_DISABLE_IMPORT_OTHER_PREFIX env var).
  * file of installed package are checked against their expected hash upon
    installation.
  * fix cache cleanup in mmpack.

 -- Nicolas Bourdaud <nicolas.bourdaud@gmail.com>  Tue, 24 Jan 2023 00:47:38 +0100

mmpack (0.7.7) unstable; urgency=medium

  * Various bugfixes.
  * Fix pycompile of /usr/share/mmpack/pyscripts.
  * Reorganize installed data and scripts in /usr/share/mmpack.
  * astroid and typing_extensions python packages are now vendored in
    mmpack-build.

 -- Nicolas Bourdaud <nicolas.bourdaud@gmail.com>  Fri, 18 Nov 2022 06:51:39 +0100

mmpack (0.7.6) unstable; urgency=medium

  * When guessing build system, prefer any other system over Makefile.
  * Add update-provides and create-ghost-specs sub command to mmpack-build
    guess.

 -- Nicolas Bourdaud <nicolas.bourdaud@gmail.com>  Wed, 12 Oct 2022 22:41:19 +0200

mmpack (0.7.5) unstable; urgency=medium

  * Support file() directive in mmpack/specs files.
  * Support token expansion in mmpack/sources-strap.
  * Add --stop-on-error option to mmpack-build mksource (and pkg-create).

 -- Nicolas Bourdaud <nicolas.bourdaud@gmail.com>  Thu, 18 Aug 2022 20:48:48 +0200

mmpack (0.7.4) unstable; urgency=medium

  * Fix build-python (wheel pkg not installed)

 -- Nicolas Bourdaud <nicolas.bourdaud@gmail.com>  Thu, 04 Aug 2022 22:45:07 +0200

mmpack (0.7.3) unstable; urgency=medium

  * Support git submodules when creating sources.
  * Add --no-prefix-mount option to mmpack run.

 -- Nicolas Bourdaud <nicolas.bourdaud@gmail.com>  Wed, 03 Aug 2022 14:55:37 +0200

mmpack (0.7.2) unstable; urgency=medium

  * Fix mmpack-build builddep.
  * Fix use of packaged mount prefix on Linux.

 -- Nicolas Bourdaud <nicolas.bourdaud@gmail.com>  Wed, 06 Jul 2022 11:54:40 +0200

mmpack (0.7.1) unstable; urgency=medium

  * fix mmpack.postinst for compat level 13.

 -- Nicolas Bourdaud <nicolas.bourdaud@gmail.com>  Mon, 04 Jul 2022 20:48:32 +0200

mmpack (0.7.0) unstable; urgency=medium

  * mmpack-build use pip to build from python source packages.
  * mmpack-build use python3 pkg argcomplete for shell completion.
  * mmpack-modifyrepo gets shell completion.
  * debhelper compat level set to 13.
  * build debian package using meson.
  * Build package using bindnow.
  * Specify Rules-Requires-Root as no.
  * Provides common licenses as symlinks to /usr/share/common-licenses.
  * extract debug symbol and package them separately.
  * execute build scripts under mmpack prefix.
  * improve shell completion of mmpack-build.
  * mmpack-build allow one to specify mmpack repo to resolve build depends.

 -- Nicolas Bourdaud <nicolas.bourdaud@gmail.com>  Mon, 04 Jul 2022 16:43:59 +0200

mmpack (0.6.10) unstable; urgency=medium

  * fix mmpack install sigabort when installing many packages.
  * fix mmpack remove segfault when removing many packages.

 -- Nicolas Bourdaud <nicolas.bourdaud@gmail.com>  Fri, 06 Aug 2021 10:15:40 +0200

mmpack (0.6.9) unstable; urgency=medium

  * PKG_CONFIG_PATH is no longer posixified with run cmd on Windows.
  * Install a site.py in prefix python site-packages upon prefix creation.
  * prefix site.py ensures that libraries installed in prefix are used when
    using python packages (with mmpack run).
  * Add --upgrade subcommand in mkprefix.

 -- Nicolas Bourdaud <nicolas.bourdaud@gmail.com>  Wed, 28 Jul 2021 11:43:20 +0200

mmpack (0.6.8) unstable; urgency=medium

  * Use regex to update version from git.
  * python compiled extension are used to enrich provides and list in the
    local files for dependency search.
  * Update PKG_CONFIG_PATH when running mmpack run.

 -- Nicolas Bourdaud <nicolas.bourdaud@gmail.com>  Thu, 24 Jun 2021 08:14:04 +0200

mmpack (0.6.7) unstable; urgency=medium

  * Support specs in utf-8 on windows.
  * Build scripts filename have the suffix "_hook" dropped. The version with
    suffix is still usable but deprecated.
  * Support for local_install build script.
  * MMPACK_BUILD_KEEP_TMPDIR env allow one to keep temporary source build
    folder.
  * Various improvements of error reporting of mmpack-build.

 -- Nicolas Bourdaud <nicolas.bourdaud@gmail.com>  Thu, 17 Jun 2021 08:12:20 +0200

mmpack (0.6.6) unstable; urgency=medium

  * Workaround python-astroid bug of inconsistent mro error.

 -- Nicolas Bourdaud <nicolas.bourdaud@gmail.com>  Fri, 11 Jun 2021 15:50:38 +0200

mmpack (0.6.5) unstable; urgency=medium

  * Improve python depends.
  * Supports 'name (>= version)' format in specs of dependencies of custom
    packages.
  * mmpack-modifyrepo support REMOVE_MATCHING_SRC batch command.
  * mmpack-modifyrepo support remove-src command.

 -- Nicolas Bourdaud <nicolas.bourdaud@gmail.com>  Thu, 10 Jun 2021 13:05:45 +0200

mmpack (0.6.4) unstable; urgency=medium

  * Only warn (instead of failure) if used shared library symbols could not be
    found in any libraries (likely a plugin on ELF platform).
  * Name mmpack package of python pkg after the import name in python. If
    needed split python package over each its top dir.
  * python -nspkg.pth considered as python package metadata.
  * Add support for mmpack-build mksource hooks.

 -- Nicolas Bourdaud <nicolas.bourdaud@gmail.com>  Wed, 09 Jun 2021 08:19:36 +0200

mmpack (0.6.3) unstable; urgency=medium

  * default prefixes and packages dir are now in XDG_CACHE_HOME/mmpack/prefix
    and XDG_CACHE_HOME/mmpack/packages.
  * Fix dependency constraints generation when writing pkginfo file.

src/sha256.c
/*********************************************************************
* Filename:   sha256.c
* Author:     Brad Conte (brad AT bradconte.com)
* Copyright:
* Disclaimer: This code is presented "as is" without any guarantees.
* Details:    Implementation of the SHA-256 hashing algorithm.
              SHA-256 is one of the three algorithms in the SHA2
              specification. The others, SHA-384 and SHA-512, are not
              offered in this implementation.
              Algorithm specification can be found here:
               * http://csrc.nist.gov/publications/fips/fips180-2/fips180-2withchangenotice.pdf
              This implementation uses little endian byte order.
*********************************************************************/
#if defined (HAVE_CONFIG_H)
# include <config.h>
#endif

/*************************** HEADER FILES ***************************/
#include <stdlib.h>
#include <stdint.h>
#include <string.h>
#include "sha256.h"

/****************************** MACROS ******************************/
#define ROTLEFT(a,b) (((a) << (b)) | ((a) >> (32-(b))))
#define ROTRIGHT(a,b) (((a) >> (b)) | ((a) << (32-(b))))

#define CH(x,y,z) (((x) & (y)) ^ (~(x) & (z)))
#define MAJ(x,y,z) (((x) & (y)) ^ ((x) & (z)) ^ ((y) & (z)))
#define EP0(x) (ROTRIGHT(x,2) ^ ROTRIGHT(x,13) ^ ROTRIGHT(x,22))
#define EP1(x) (ROTRIGHT(x,6) ^ ROTRIGHT(x,11) ^ ROTRIGHT(x,25))
#define SIG0(x) (ROTRIGHT(x,7) ^ ROTRIGHT(x,18) ^ ((x) >> 3))
#define SIG1(x) (ROTRIGHT(x,17) ^ ROTRIGHT(x,19) ^ ((x) >> 10))

/**************************** VARIABLES *****************************/
static const uint32_t k[64] = {
	0x428a2f98,0x71374491,0xb5c0fbcf,0xe9b5dba5,0x3956c25b,0x59f111f1,0x923f82a4,0xab1c5ed5,
	0xd807aa98,0x12835b01,0x243185be,0x550c7dc3,0x72be5d74,0x80deb1fe,0x9bdc06a7,0xc19bf174,
	0xe49b69c1,0xefbe4786,0x0fc19dc6,0x240ca1cc,0x2de92c6f,0x4a7484aa,0x5cb0a9dc,0x76f988da,
	0x983e5152,0xa831c66d,0xb00327c8,0xbf597fc7,0xc6e00bf3,0xd5a79147,0x06ca6351,0x14292967,
	0x27b70a85,0x2e1b2138,0x4d2c6dfc,0x53380d13,0x650a7354,0x766a0abb,0x81c2c92e,0x92722c85,
	0xa2bfe8a1,0xa81a664b,0xc24b8b70,0xc76c51a3,0xd192e819,0xd6990624,0xf40e3585,0x106aa070,
	0x19a4c116,0x1e376c08,0x2748774c,0x34b0bcb5,0x391c0cb3,0x4ed8aa4a,0x5b9cca4f,0x682e6ff3,
	0x748f82ee,0x78a5636f,0x84c87814,0x8cc70208,0x90befffa,0xa4506ceb,0xbef9a3f7,0xc67178f2
};

/*********************** FUNCTION DEFINITIONS ***********************/

static
void sha256_transform(SHA256_CTX *ctx, const unsigned char* data)
{
	uint32_t a, b, c, d, e, f, g, h, t1, t2, m[64];
	int i, j;

	for (i = 0, j = 0; i < 16; ++i, j += 4)
		m[i] = (data[j] << 24) | (data[j + 1] << 16) | (data[j + 2] << 8) | (data[j + 3]);
	for ( ; i < 64; ++i)
		m[i] = SIG1(m[i - 2]) + m[i - 7] + SIG0(m[i - 15]) + m[i - 16];

	a = ctx->state[0];
	b = ctx->state[1];
	c = ctx->state[2];
	d = ctx->state[3];
	e = ctx->state[4];
	f = ctx->state[5];
	g = ctx->state[6];
	h = ctx->state[7];

	for (i = 0; i < 64; ++i) {
		t1 = h + EP1(e) + CH(e,f,g) + k[i] + m[i];
		t2 = EP0(a) + MAJ(a,b,c);
		h = g;
		g = f;
		f = e;
		e = d + t1;
		d = c;
		c = b;
		b = a;
		a = t1 + t2;
	}

	ctx->state[0] += a;
	ctx->state[1] += b;
	ctx->state[2] += c;
	ctx->state[3] += d;
	ctx->state[4] += e;
	ctx->state[5] += f;
	ctx->state[6] += g;
	ctx->state[7] += h;
}


LOCAL_SYMBOL
void sha256_init(SHA256_CTX *ctx)
{
	ctx->datalen = 0;
	ctx->bitlen = 0;
	ctx->state[0] = 0x6a09e667;
	ctx->state[1] = 0xbb67ae85;
	ctx->state[2] = 0x3c6ef372;
	ctx->state[3] = 0xa54ff53a;
	ctx->state[4] = 0x510e527f;
	ctx->state[5] = 0x9b05688c;
	ctx->state[6] = 0x1f83d9ab;
	ctx->state[7] = 0x5be0cd19;
}


LOCAL_SYMBOL
void sha256_update(SHA256_CTX *ctx, const void* buffer, size_t len)
{
	size_t i;
	const unsigned char* restrict data = buffer;

	for (i = 0; i < len; ++i) {
		ctx->data[ctx->datalen] = data[i];
		ctx->datalen++;
		if (ctx->datalen == 64) {
			sha256_transform(ctx, ctx->data);
			ctx->bitlen += 512;
			ctx->datalen = 0;
		}
	}
}


LOCAL_SYMBOL
void sha256_final(SHA256_CTX *ctx, unsigned char* hash)
{
	int i;

	i = ctx->datalen;

	// Pad whatever data is left in the buffer.
	if (ctx->datalen < 56) {
		ctx->data[i++] = 0x80;
		while (i < 56)
			ctx->data[i++] = 0x00;
	}
	else {
		ctx->data[i++] = 0x80;
		while (i < 64)
			ctx->data[i++] = 0x00;
		sha256_transform(ctx, ctx->data);
		memset(ctx->data, 0, 56);
	}

	// Append to the padding the total message's length in bits and transform.
	ctx->bitlen += ctx->datalen * 8;
	ctx->data[63] = ctx->bitlen;
	ctx->data[62] = ctx->bitlen >> 8;
	ctx->data[61] = ctx->bitlen >> 16;
	ctx->data[60] = ctx->bitlen >> 24;
	ctx->data[59] = ctx->bitlen >> 32;
	ctx->data[58] = ctx->bitlen >> 40;
	ctx->data[57] = ctx->bitlen >> 48;
	ctx->data[56] = ctx->bitlen >> 56;
	sha256_transform(ctx, ctx->data);

	// Since this implementation uses little endian byte ordering and SHA uses big endian,
	// reverse all the bytes when copying the final state to the output hash.
	for (i = 0; i < 4; ++i) {
		hash[i]      = (ctx->state[0] >> (24 - i * 8)) & 0x000000ff;
		hash[i + 4]  = (ctx->state[1] >> (24 - i * 8)) & 0x000000ff;
		hash[i + 8]  = (ctx->state[2] >> (24 - i * 8)) & 0x000000ff;
		hash[i + 12] = (ctx->state[3] >> (24 - i * 8)) & 0x000000ff;
		hash[i + 16] = (ctx->state[4] >> (24 - i * 8)) & 0x000000ff;
		hash[i + 20] = (ctx->state[5] >> (24 - i * 8)) & 0x000000ff;
		hash[i + 24] = (ctx->state[6] >> (24 - i * 8)) & 0x000000ff;
		hash[i + 28] = (ctx->state[7] >> (24 - i * 8)) & 0x000000ff;
	}
}
